packages/ozone/src/daemon/blob-diverter.ts

```diff
@@ -0,0 +1,229 @@
+import axios from 'axios'
+import { Readable } from 'stream'
+import { CID } from 'multiformats/cid'
+
+import Database from '../db'
+import { retryHttp } from '../util'
+import { dbLogger } from '../logger'
+import { BlobReportServiceConfig } from '../config'
+
+type PollState = {
+  timer?: NodeJS.Timer
+  promise: Promise<void>
+}
+
+export class BlobDiverter {
+  destroyed = false
+
+  pollState: PollState = {
+    promise: Promise.resolve(),
+  }
+
+  serviceConfig: BlobReportServiceConfig
+  idResolver: IdResolver
+
+  constructor(
+    public db: Database,
+    services: {
+      idResolver: IdResolver
+      serviceConfig: BlobReportServiceConfig
+    },
+  ) {
+    this.serviceConfig = services.serviceConfig
+    this.idResolver = services.idResolver
+  }
+
+  start() {
+    this.poll(this.pollState, () => this.divertBlob())
+  }
+
+  poll(state: PollState, fn: () => Promise<void>) {
+    if (this.destroyed) return
+    state.promise = fn()
+      .catch((err) => {
+        dbLogger.error({ err }, 'blob divert failed')
+      })
+      .finally(() => {
+        state.timer = setTimeout(() => this.poll(state, fn), 30 * SECOND)
+      })
+  }
+
+  async processAll() {
+    await Promise.all([this.divertBlob(), this.pollState.promise])
+  }
+
+  async destroy() {
+    this.destroyed = true
+    const destroyState = (state: PollState) => {
+      if (state.timer) {
+        clearTimeout(state.timer)
+      }
+      return state.promise
+    }
+    await destroyState(this.pollState)
+  }
+
+  async divertBlob() {
+    const toPush = await this.db.db
+      .selectFrom('blob_divert_event')
+      .select('id')
+      .forUpdate()
+      .skipLocked()
+      .where('divertedAt', 'is', null)
+      .where('attempts', '<', 10)
+      .execute()
+    await Promise.all(toPush.map((evt) => this.attemptBlobDivert(evt.id)))
+  }
+
+  private async getBlob({
+    pds,
+    did,
+    cid,
+  }: {
+    pds: string
+    did: string
+    cid: string
+  }) {
+    const blobResponse = await axios.get(
+      `${pds}/xrpc/com.atproto.sync.getBlob`,
+      {
+        params: { did, cid },
+        decompress: true,
+        responseType: 'stream',
+        timeout: 5000, // 5sec of inactivity on the connection
+      },
+    )
+    const imageStream: Readable = blobResponse.data
+    const verifyCid = new VerifyCidTransform(CID.parse(cid))
+    forwardStreamErrors(imageStream, verifyCid)
+
+    return {
+      contentType:
+        blobResponse.headers['content-type'] || 'application/octet-stream',
+      imageStream: imageStream.pipe(verifyCid),
+    }
+  }
+
+  private async uploadBlob(
+    {
+      imageStream,
+      contentType,
+    }: { imageStream: Readable; contentType: string },
+    { subjectDid, subjectUri }: { subjectDid: string; subjectUri: string },
+  ) {
+    if (!this.serviceConfig.authToken || !this.serviceConfig.url) {
+      return false
+    }
+
+    const url = `${this.serviceConfig.url}?did=${subjectDid}&uri=${subjectUri}`
```

Comment:
We will want to ensure these parts get %-encoded.  The most reliable way to do it is probably to build a `URL`.:
```js
const url = new URL(this.serviceConfig.url)
url.searchParams.set('did', subjectDid)
url.searchParams.set('uri', subjectUri)
```